File: JAVA/(DP)Partition Equal Subset Sum.java
//https://leetcode.com/problems/partition-equal-subset-sum/


class Solution {
    public boolean canPartition(int[] nums) {
        int len = nums.length;
        if(len<2)   return false;
        int sum = 0;
        for(int val:nums)   sum+=val;
        if(sum%2==1)    return false;
        int target = sum/2;
        boolean[] dp = new boolean[target+1];
        dp[0] = true;
        for(int i = 0 ; i<len ; i++){
            for(int j = target; j>=nums[i]; j--){
                dp[j] = dp[j] || dp[j-nums[i]];
            }
        }
        return dp[target];
    }
}

File: go/(DP)Partition Equal Subset Sum.go
//https://leetcode.com/problems/partition-equal-subset-sum/


func canPartition(nums []int) bool {
    n:=len(nums)
    if n==1 {return false}
    sum:=0
    for _,val := range nums{
        sum+=val
    }
    if sum%2==1 {return false}
    sum/=2
    dp := make([]bool, sum+1)
    dp[0] = true
    for _,val := range nums {
        for i:=sum ; i>=val ; i-- {
            dp[i] = dp[i]||dp[i-val]
        }
    }
    return dp[sum]
}